### ISSUE TITLE
Using <local> for Bypass List prevents IPv6 from working

### ISSUE THREAD

**Issue by electrum** [FelisCatus/SwitchyOmega#443]
The default `<local>` bypass entry seems to prevent any IPv6 connections from going through the proxy:

    http://[2607:f8b0:400a:807::200e]

This is likely a bug in Chrome: https://code.google.com/p/chromium/issues/detail?id=475521

**Comment by FelisCatus** [comment]
I'll just keep this issue open for the record.

**Comment by FelisCatus** [comment]
This is confirmed by another report and is likely to be an issue with Chromium. Please provide additional information to the linked issue above.

Meanwhile, shall we just change `<local>` to `127.0.0.1`, `localhost` and `[::1]` internally in SwitchyOmega to work around the problem? SwitchyOmega does have a chance to do so before handling the proxy settings to Chromium, but I don't know whether this is the right thing to do.

**Comment by FelisCatus** [comment]
In my opinion, the issue won't be fixed soon in Chromium and can take some time before it hits stable. Besides, there are no unwanted side-effects to apply the mapping, expect switching to a proxy profile becomes just a little bit slower.

So let's just do the hack. I just mapped `<local>` to `127.0.0.1`, `localhost` and `[::1]` in the source code and hope everything can be fixed. The new release containing the hack is scheduled on this weekend.

**Comment by FelisCatus** [comment]
The fix was just released with v2.3.14.

**Comment by FelisCatus** [comment]
Hi, it turns out that `<local>` is not the same as `127.0.0.1`, `localhost` and `[::1]`.

 [Chromium considers](https://code.google.com/p/chromium/codesearch#chromium/src/net/proxy/proxy_bypass_rules.cc&sq=package:chromium&l=65) a host is `<local>` if `(host == "127.0.0.1" || host == "[::1]")`, **or host does not contain the character `.`(dot).** It has nothing to do with `localhost`.

However, [the Chrome extension documentation](https://developer.chrome.com/extensions/proxy#bypass_list) contains misleading information, quote:

> Match local addresses. An address is local if the host is "127.0.0.1", "::1", or "localhost".

Therefore I was tricked into thinking that I can work around the issue by doing the replacement. I was wrong. So was the extension docs.

----

**tl;dr** The "fix" changed the original behavior of `<local>` in SwitchyOmega, and it has caused troubles like #683.

The misunderstanding of `<local>` has also influenced other concepts and implementations in SwitchyOmega, like PAC script generation for `<local>` and various UI elements.

This must be fixed for real. The most sensible way is to keep `<local>` as-is. However, that would also bring the Chromium bug back. What shall we do? What works better for users?

**Comment by FelisCatus** [comment]
Sorry, but I cannot provide any useful links related to this issue.

I tried to search the Chromium issue tracker, but I was overwhelmed by the large amount of results and could not find any relevant issues. I also googled for `chrome/chromium proxy "<local>"`, but apparently none of these words are good keywords.

**Comment by FelisCatus** [comment]
Personally I want to revert the `<local>` mapping here to align with standard behavior in system profile.

For users who are already using `<local>`, we can automatically change it to these three entries: `127.0.0.1`, `localhost` and `[::1]`. And for new users who manually type `<local>` there, it wouldn't be confusing that the same `<local>` rule works differently in system settings or in SwitchyOmega.

One more question though: How could you bypass all one-level domains but not all IPv6 literals? It doesn't seem to be possible without using `<local>` and fall into that bug. It could be possible with Switch Profiles and Host Levels Conditions, but...

**Comment by FelisCatus** [comment]
I would also want to add help text to clarify what `<local>` really does. But that would be...

> `<local>` will match `127.0.0.1`, `[::1]`, and any host which doesn't contain a dot (`.`). **Be careful though, because all IPv6 literals will also be affected due to lack of dots.**

That doesn't sound great to me. Shall I append "blame Chrome for this"? :no_good:

**Comment by FelisCatus** [comment]
Or shall we keep the (technically incorrect) existing `<local>` behavior and add something like `<local-and-hosts-without-dots-also-ipv6-address-literals-DAMMIT>`, which translates to `<local>` in Chrome? :wink:

**Comment by FelisCatus** [comment]
Sorry, but I'm reverting the changes now and pass `<local>` as-is. `<local>` will match any host without dots in both Chrome and Firefox, and I think it is the intended behavior.

As for IPv6 literals unfortunately affected by `<local>`, I consider them to be caused by design flaws in Chrome and Firefox. However, attempting to "fix" that on SwitchyOmega's side causes problems and confusion. For example, the users lose the capability of bypassing all intranet domains (a.k.a. domains without dots). And it is hard for anyone to figure out why. (See #683 for the story.)

For the misleading documentation on Chrome extensions, I'd better open a issue somewhere when I have time for it.

**Comment by FelisCatus** [comment]
For users like @electrum: Please manually change `<local>` to `127.0.0.1`, `localhost` and `[::1]` if that behavior is desired.

**Comment by FelisCatus** [comment]
The changes kicked in as of v2.3.18.

* The behavior of `<local>` in Bypass List has changed to follow the browser and system.
  - If `<local>` appears in the list, `127.0.0.1`, `[::1]` and any other host NOT containing dots(`.`) will be bypassed.
  - **Attention: IPv6 literals are also bypassed because they don't have dots in them. This is a limitation of the browser and cannot be fixed.**
